File: nodiguard/shields/NodiGuardShield.kt
package com.nodiguard.shield

import android.content.Context
import android.content.pm.PackageManager
import android.os.Debug
import android.os.Process
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.net.HttpURLConnection
import java.net.Socket
import java.net.URL
import java.security.MessageDigest
import kotlin.concurrent.thread

/**
 * NodiGuard Active RASP (Runtime Application Self-Protection) Wall for Android.
 * Neutralizes APK decompilation, Smali tampering, Frida dynamic hooks, and Magisk root injection.
 */
object NodiGuardShield {

    // HONEYPOT TRAP: Decompiled strings in Jadx/Apktool will expose this honey-token.
    // If an attacker sends any request with this token, Cloudflare WAF bans their IP permanently.
    const val CANARY_HONEY_TOKEN = "paintrace_canary_honey_trap_android_c729ab"

    fun armWall(context: Context, gatewayUrl: String = "https://ai.iotservices.my", expectedCertSha256: String? = null) {
        thread(isDaemon = true) {
            detectRoot(context, gatewayUrl)
            detectFrida(gatewayUrl)
            detectDebugger(gatewayUrl)
            detectProxy(gatewayUrl)
            if (expectedCertSha256 != null) {
                verifyApkSignature(context, expectedCertSha256, gatewayUrl)
            }
        }
    }

    /** 1. Detects Root & Magisk binaries */
    private fun detectRoot(context: Context, gatewayUrl: String) {
        val rootPaths = arrayOf(
            "/system/app/Superuser.apk",
            "/sbin/su",
            "/system/bin/su",
            "/system/xbin/su",
            "/data/local/xbin/su",
            "/data/local/bin/su",
            "/system/sd/xbin/su",
            "/system/bin/failsafe/su",
            "/data/local/su"
        )
        for (path in rootPaths) {
            if (File(path).exists()) {
                triggerTamperDefense("Root Binary Found at $path", gatewayUrl)
                return
            }
        }
    }

    /** 2. Detects Frida Dynamic Hook Server & frida-agent.so in memory */
    private fun detectFrida(gatewayUrl: String) {
        // A. Check for Frida default listening port 27042
        try {
            Socket("127.0.0.1", 27042).use {
                triggerTamperDefense("Frida Hook Server Port 27042 Active", gatewayUrl)
                return
            }
        } catch (_: Exception) {}

        // B. Inspect /proc/self/maps for injected frida-agent or xposed libraries
        try {
            val reader = BufferedReader(FileReader("/proc/self/maps"))
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                if (line!!.contains("frida-agent") || line!!.contains("gadget.so") || line!!.contains("xposed")) {
                    reader.close()
                    triggerTamperDefense("Injected Hook Library Found in Memory: $line", gatewayUrl)
                    return
                }
            }
            reader.close()
        } catch (_: Exception) {}
    }

    /** 3. Detects ADB / Android Studio Debugger */
    private fun detectDebugger(gatewayUrl: String) {
        if (Debug.isDebuggerConnected() || Debug.waitingForDebugger()) {
            triggerTamperDefense("Active Debugger Connected to Android Process", gatewayUrl)
        }
    }

    /** 4. Detects Charles / Fiddler MITM Proxy */
    private fun detectProxy(gatewayUrl: String) {
        val proxyHost = System.getProperty("http.proxyHost")
        val proxyPort = System.getProperty("http.proxyPort")
        if (!proxyHost.isNullOrEmpty() && !proxyPort.isNullOrEmpty()) {
            triggerTamperDefense("MITM Proxy Detected ($proxyHost:$proxyPort)", gatewayUrl)
        }
    }

    /** 5. Verifies APK signature hash against tampering & re-signing */
    private fun verifyApkSignature(context: Context, expectedSha256: String, gatewayUrl: String) {
        try {
            val packageInfo = context.packageManager.getPackageInfo(
                context.packageName,
                PackageManager.GET_SIGNATURES
            )
            for (sig in packageInfo.signatures) {
                val md = MessageDigest.getInstance("SHA-256")
                val digest = md.digest(sig.toByteArray())
                val hashHex = digest.joinToString("") { "%02x".format(it) }
                if (!hashHex.equals(expectedSha256, ignoreCase = true)) {
                    triggerTamperDefense("APK Signature Mismatch! Repackaging Detected.", gatewayUrl)
                    return
                }
            }
        } catch (_: Exception) {}
    }

    /** 6. Reaction: Dispatches telemetry beacon to NodiGuard / Cloudflare, then terminates process */
    private fun triggerTamperDefense(reason: String, gatewayUrl: String) {
        thread {
            try {
                val url = URL("$gatewayUrl/api/security/tamper-alert")
                val conn = url.openConnection() as HttpURLConnection
                conn.requestMethod = "POST"
                conn.setRequestProperty("Content-Type", "application/json")
                conn.doOutput = true
                conn.connectTimeout = 3000
                conn.readTimeout = 3000

                val payload = """
                    {"platform": "Android", "reason": "$reason"}
                """.trimIndent()

                conn.outputStream.use { it.write(payload.toByteArray()) }
                conn.responseCode
            } catch (_: Exception) {}

            // Immediately kill process
            Process.killProcess(Process.myPid())
            System.exit(0)
        }
    }
}

File: nodiguard/shields/NodiGuardShield.swift
//
//  NodiGuardShield.swift
//  NodiGuard Active RASP (Runtime Application Self-Protection) Wall for iOS.
//  Protects against IPA decompilation, Frida dynamic hooks, LLDB debugging, and Jailbreak tampering.
//

import Foundation
import UIKit
import Darwin

public final class NodiGuardShield {
    public static let shared = NodiGuardShield()
    
    // HONEYPOT TRAP: Decompiled strings will reveal this decoy token.
    // If an attacker attempts to query the API with this honey-token, their IP is immediately banned on Cloudflare WAF.
    public let canaryHoneyToken = "paintrace_canary_honey_trap_ios_9981aef0"
    
    private init() {}
    
    /// Initializes and arms all anti-reverse engineering sensors.
    public func armWall(gatewayURL: String = "https://ai.iotservices.my") {
        #if !DEBUG
        checkJailbreak(gatewayURL: gatewayURL)
        detectDebugger(gatewayURL: gatewayURL)
        detectFridaHook(gatewayURL: gatewayURL)
        preventDebuggerAttachment()
        #endif
    }
    
    /// 1. Detects Jailbreak environment (Cydia, Sileo, Substrate, sandbox escape)
    private func checkJailbreak(gatewayURL: String) {
        let suspiciousPaths = [
            "/Applications/Cydia.app",
            "/Library/MobileSubstrate/MobileSubstrate.dylib",
            "/bin/bash",
            "/usr/sbin/sshd",
            "/etc/apt",
            "/usr/bin/ssh",
            "/private/var/lib/apt"
        ]
        for path in suspiciousPaths {
            if FileManager.default.fileExists(atPath: path) {
                triggerTamperDefense(reason: "Jailbreak Binary Detected at \(path)", gatewayURL: gatewayURL)
                return
            }
        }
        
        // Sandbox write test
        let testPath = "/private/nodiguard_jailbreak_test.txt"
        do {
            try "tamper_test".write(toFile: testPath, atomically: true, encoding: .utf8)
            try? FileManager.default.removeItem(atPath: testPath)
            triggerTamperDefense(reason: "Sandbox Integrity Violated (Root Write Permitted)", gatewayURL: gatewayURL)
        } catch {
            // Normal sandboxed behavior
        }
    }
    
    /// 2. Detects LLDB / GDB debugger attachment via sysctl
    private func detectDebugger(gatewayURL: String) {
        var info = kinfo_proc()
        var mib: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()]
        var size = MemoryLayout<kinfo_proc>.stride
        let junk = sysctl(&mib, UInt32(mib.count), &info, &size, nil, 0)
        if junk == 0 && (info.kp_proc.p_flag & P_TRACED) != 0 {
            triggerTamperDefense(reason: "Active Debugger Attached (P_TRACED Flag Set)", gatewayURL: gatewayURL)
        }
    }
    
    /// 3. Detects Frida Dynamic Instrumentation Server & Dylibs
    private func detectFridaHook(gatewayURL: String) {
        // Check for Frida default listening port 27042
        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = in_port_t(27042).bigEndian
        addr.sin_addr.s_addr = inet_addr("127.0.0.1")
        
        let sock = socket(AF_INET, SOCK_STREAM, 0)
        if sock >= 0 {
            let res = withUnsafePointer(to: &addr) {
                $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                    connect(sock, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
                }
            }
            close(sock)
            if res == 0 {
                triggerTamperDefense(reason: "Frida Server Port 27042 Open (Active Hooking)", gatewayURL: gatewayURL)
                return
            }
        }
    }
    
    /// 4. Disables ptrace attachment at kernel level
    private func preventDebuggerAttachment() {
        let ptracePtr = dlsym(dlopen(nil, RTLD_NOW), "ptrace")
        if let ptracePtr = ptracePtr {
            typealias PtraceType = @convention(c) (CInt, pid_t, CInt, CInt) -> CInt
            let ptraceFunc = unsafeBitCast(ptracePtr, to: PtraceType.self)
            _ = ptraceFunc(31, 0, 0, 0) // PT_DENY_ATTACH = 31
        }
    }
    
    /// 5. Self-Defense Reaction: Dispatches telemetry beacon and self-destructs
    private func triggerTamperDefense(reason: String, gatewayURL: String) {
        // Dispatches telemetry beacon to trigger automated Cloudflare IP Ban
        guard let url = URL(string: "\(gatewayURL)/api/security/tamper-alert") else {
            exit(0)
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let payload: [String: Any] = [
            "platform": "iOS",
            "reason": reason,
            "device": UIDevice.current.model,
            "timestamp": ISO8601DateFormatter().string(from: Date())
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)
        
        let task = URLSession.shared.dataTask(with: request) { _, _, _ in
            // Exit immediately after sending threat beacon
            exit(0)
        }
        task.resume()
        
        // Block main thread to prevent further execution
        Thread.sleep(forTimeInterval: 0.5)
        exit(0)
    }
}
